lib/puppet_x/puppetlabs/cd4pe_job_helper.rb

```diff
@@ -0,0 +1,166 @@
+          ENV[key] = value
+      end
+    end
+
+    def make_working_dir(working_dir)
+      Dir.mkdir(working_dir) unless File.exists?(working_dir)
+    end
+
+    def get_job_script_and_control_repo
+      target_file = File.join(@working_dir, "cd4pe_job.tar.gz")
+      client = PuppetX::Puppetlabs::CD4PEClient.new
+      response = client.get_job_script_and_control_repo(target_file)
+      case response
+      when Net::HTTPNotFound
+        raise "Message: #{response.body}\nCode: #{response.code}"
+      when Net::HTTPServerError
+        raise "Unknown HTTP Error with code: #{response.code} and body #{response.body}"
+      end
+
+      target_file
+    end
+
+    def run_job
+      result = execute_manifest(MANIFEST_TYPE[:JOB])
+      
+      if (result[:exit_code] == 0)
+        on_job_success(result)
+      else
+        on_job_failure(result)
+      end
+    end
+
+    def on_job_success(result)
+      output = {}
+      output[:job] = {
+        exit_code: result[:exit_code], 
+        message: result[:message]
+      }
+    
+      # if a AFTER_JOB_SUCCESS script exists, run it now!
+      run_after_success = File.exists?(File.join(@local_jobs_dir, MANIFEST_TYPE[:AFTER_JOB_SUCCESS]))
+      if (run_after_success)
+        @@logs.push("#{MANIFEST_TYPE[:AFTER_JOB_SUCCESS]} script specified.")
+        success_script_result = execute_manifest(MANIFEST_TYPE[:AFTER_JOB_SUCCESS])
+        output[:after_job_success] = {
+          exit_code: success_script_result[:exit_code], 
+          message: success_script_result[:message]
+        }
+      end
+    
+      output
+    end
+
+    def on_job_failure(result)
+      output = {}
+      output[:job] = {
+        exit_code: result[:exit_code], 
+        message: result[:message]
+      }
+    
+      # if a AFTER_JOB_FAILURE script exists, run it now!
+      run_after_failure = File.exists?(File.join(@local_jobs_dir, MANIFEST_TYPE[:AFTER_JOB_FAILURE]))
+      if (run_after_failure)
+        @@logs.push("#{MANIFEST_TYPE[:AFTER_JOB_FAILURE]} script specified.")
+        failure_script_result = execute_manifest(MANIFEST_TYPE[:AFTER_JOB_FAILURE])
+        output[:after_job_failure] = {
+          exit_code: failure_script_result[:exit_code], 
+          message: failure_script_result[:message]
+        }
+      end
+
+      output
+    end
+
+    def execute_manifest(manifest_type)
+      @@logs.push("Executing #{manifest_type} manifest.")
+      result = {}
+      if (@docker_based_job)
+        @@logs.push("Docker image specified. Running #{manifest_type} manifest on docker image: #{@docker_image}.")
+        result = run_with_docker(manifest_type)
+      else
+        @@logs.push("No docker image specified. Running #{manifest_type} manifest directly on machine.")
+        result = run_with_system(manifest_type)
+      end
+      
+      if (result[:exit_code] == 0)
+        @@logs.push("#{manifest_type} succeeded!")
+      else 
+        @@logs.push("#{manifest_type} failed!")
```

Comment:
Wondering if we should print out the exit code, because if it fails the exit code is useful to know *why*.

Reply:
We DO return the exit codes as part of the Bolt json return payload, but I agree it would be helpful in the logs as well. Adding!